examples/shop/src/store/bucket.tsx

```diff
@@ -0,0 +1,58 @@
+import React, { createContext, useContext } from 'react';
+import { Product } from '../types';
+
+export type AddedProduct = Product & { count: number };
+
+export type AddedProducts = Map<number, AddedProduct>;
+
+type BucketContextType = {
+    addedProducts: AddedProducts;
+    addProduct: (product: Product) => void;
+    removeProduct: (product: Product) => void;
+};
+
+const BucketContext = createContext<BucketContextType>({} as BucketContextType);
+
+type BucketProviderProps = {
+    children: React.ReactNode;
+};
+
+export function BucketProvider({ children }: BucketProviderProps) {
+    const [addedProducts, setAddedProducts] = React.useState<AddedProducts>(new Map());
+
+    function addProduct(newProduct: Product) {
+        setAddedProducts((addedProducts) => {
+            if (addedProducts.has(newProduct.id)) {
+                const addedProduct = addedProducts.get(newProduct.id) as AddedProduct;
+
+                addedProduct.count += 1;
+            } else {
+                addedProducts.set(newProduct.id, { ...newProduct, count: 1 });
```

Comment:
Please, do not mutate `addedProducts` if you anyway return new 'Map'. This can lead to side effects in the future. (see below)

Reply:
fixed